src/dfx/src/commands/canister/request_status.rs

```diff
@@ -46,46 +48,61 @@ pub async fn exec(env: &dyn Environment, opts: RequestStatusOpts) -> DfxResult {
     let canister_id = Principal::from_text(callee_canister)
         .or_else(|_| canister_id_store.get(callee_canister))?;
 
-    let mut retry_policy = ExponentialBackoff::default();
-    let blob = async {
-        let mut request_accepted = false;
-        loop {
-            let (response, _cert) = agent
-                .request_status_raw(&request_id, canister_id)
-                .await
-                .context("Failed to fetch request status.")?;
-            match response {
-                RequestStatusResponse::Replied(reply) => return Ok(reply.arg),
-                RequestStatusResponse::Rejected(response) => {
-                    return Err(DfxError::new(AgentError::CertifiedReject(response)))
-                }
-                RequestStatusResponse::Unknown => (),
-                RequestStatusResponse::Received | RequestStatusResponse::Processing => {
-                    // The system will return Unknown until the request is accepted
-                    // and we generally cannot know how long that will take.
-                    // State transitions between Received and Processing may be
-                    // instantaneous. Therefore, once we know the request is accepted,
-                    // we restart the waiter so the request does not time out.
-                    if !request_accepted {
-                        retry_policy.reset();
-                        request_accepted = true;
+    let blob = if let Some(pocketic) = env.get_pocketic() {
+        let msg_id = RawMessageId {
```

Comment:
is this change related to impersonation?

Reply:
Indeed. Because request status can only be retrieved by the same caller who submitted the update call before.

Reply:
This new code path is always taken when `env.get_pocketic.is_some()`, whether or not `--impersonate` is involved. Doesn't this mean that when using PocketIC, `dfx canister request_status` will work no matter which identity is used?

I'm adding a new test in https://github.com/dfinity/sdk/pull/4047 to demonstrate this.

Reply:
> Doesn't this mean that when using PocketIC, dfx canister request_status will work no matter which identity is used?

Correct. If we aren't comfortable with this behavior, PocketIC would need to be extended with a check for matching callers (in production, the read state request handler takes care of it, but in PocketIC this was omitted since security is not a requirement and it didn't seem important for the sake of _canister_ testing).

Reply:
Here's how this could apply to canister testing:
- developer writes canister
- some of developer's maintenance processes include `dfx canister call --async` and `dfx canister request-status`, but they forget to pass the same `--identity` on both
- this all works locally since they're using PocketIC, but fails on mainnet